### PR TITLE
feat: Jenkins fetch coverage

### LINKED ISSUE #1033: No coverage: Implement Jenkins
**Description**
Implement fetchCoverage for Jenkins.

**Acceptance Criteria**
- [ ] Implement fetchCoverage for Jenkins with tests
- [ ] Test sync using with Jenkins limited permissions access key using dev environment

**Additional context**
[CI integrations module architecture](https://github.com/platform-platform/monorepo/blob/master/metrics/ci_integrations/docs/01_ci_integration_module_architecture.md)

### PR DESCRIPTION
**Pull request by solid-tetianamovlian**
Fixes #1033 

## Checklist

- [x] Documented
- [x] Unit tested

### CODE REVIEW

File: metrics/ci_integrations/test/source/jenkins/adapter/jenkins_source_client_adapter_test.dart [hunk 1]
```diff
@@ -683,6 +612,165 @@ void main() {
...
+            .thenSuccessWith(defaultArtifactContent);
+
+        await adapter.fetchCoverage(defaultBuild);
+
+        verifyNever(jenkinsClientMock.fetchArtifactByRelativePath(any, any));
+      },
+    );
+
+    test(
+      ".fetchCoverage() returns null if an artifact bytes are null",
```
Comment (solid-vladlaba) [inline metrics/ci_integrations/test/source/jenkins/adapter/jenkins_source_client_adapter_test.dart:737]:
```suggestion
      ".fetchCoverage() returns null if an artifact content is null",
```

File: metrics/ci_integrations/test/source/jenkins/adapter/jenkins_source_client_adapter_test.dart [hunk 2]
```diff
@@ -683,6 +612,165 @@ void main() {
...
+        expect(result, throwsStateError);
+      },
+    );
+
+    test(
+      ".fetchCoverage() throws a StateError if fetching a jenkins build fails",
+      () {
+        whenFetchCoverage(withJenkinsBuild: defaultJenkinsBuild)
+            .thenSuccessWith(defaultArtifactContent);
+
```
Comment (solid-vladlaba) [inline metrics/ci_integrations/test/source/jenkins/adapter/jenkins_source_client_adapter_test.dart:765]:
```suggestion
```

File: metrics/ci_integrations/test/source/jenkins/adapter/jenkins_source_client_adapter_test.dart [hunk 3]
```diff
@@ -683,6 +612,165 @@ void main() {
...
+        whenFetchBuilds().thenAnswer(responses.fetchBuilds);
+
+        final result = await adapter.fetchBuildsAfter(jobName, build);
+
+        final actualAPIUrl = result.first.apiUrl;
+
+        final expectedAPIUrl =
+            '${jenkinsBuild.url}${JenkinsClient.jsonApiPath}';
+
+        expect(actualAPIUrl, equals(expectedAPIUrl));
```
Comment (solid-vladlaba) [inline metrics/ci_integrations/test/source/jenkins/adapter/jenkins_source_client_adapter_test.dart:639]:
```suggestion
        const build = BuildData(buildNumber: 1);
        const jenkinsBuild = JenkinsBuild(
          number: 2,
          url: defaultBuildUrl,
          building: false,
          artifacts: [defaultArtifact],
        );
        final expectedAPIUrl =
            '${jenkinsBuild.url}${JenkinsClient.jsonApiPath}';
       
        responses.addBuilds([jenkinsBuild]);
        whenFetchBuilds().thenAnswer(responses.fetchBuilds);

        final result = await adapter.fetchBuildsAfter(jobName, build);
        final actualAPIUrl = result.first.apiUrl;

        expect(actualAPIUrl, equals(expectedAPIUrl));
```

File: metrics/ci_integrations/test/source/jenkins/adapter/jenkins_source_client_adapter_test.dart [hunk 4]
```diff
@@ -683,6 +612,165 @@ void main() {
...
+
+        responses.addBuilds([jenkinsBuild]);
+
+        whenFetchBuilds().thenAnswer(responses.fetchBuilds);
+
+        final result = await adapter.fetchBuildsAfter(jobName, build);
+
+        final actualAPIUrl = result.first.apiUrl;
+
+        expect(actualAPIUrl, isNull);
```
Comment (solid-vladlaba) [inline metrics/ci_integrations/test/source/jenkins/adapter/jenkins_source_client_adapter_test.dart:663]:
```suggestion
        const build = BuildData(buildNumber: 1);
        const jenkinsBuild = JenkinsBuild(
          number: 2,
          url: null,
          building: false,
          artifacts: [defaultArtifact],
        );

        responses.addBuilds([jenkinsBuild]);
        whenFetchBuilds().thenAnswer(responses.fetchBuilds);

        final result = await adapter.fetchBuildsAfter(jobName, build);
        final actualAPIUrl = result.first.apiUrl;

        expect(actualAPIUrl, isNull);
```

**Review by solid-vladlaba — CHANGES_REQUESTED** [review]
Well done, @solid-maxim @solid-tetianamovlian! I've added a couple of suggestions below - please let me know what you think!

**Review by illia-romanenko — APPROVED** [review]
I did a quick scan - all looks good to me - thanks @solid-tetianamovlian @solid-maxim  @solid-vladlaba !

**Review by solid-vladlaba — APPROVED** [review]
LGTM, thanks @solid-maxim @solid-tetianamovlian!:rocket: 
I've performed a couple of improvements and introduced several classes to make Jenkins integration more granular and clear. Please review them and let me know what you think!

**Pull request merged**